File: InsertionSortC/Logic.cpp
/*
 * Logic.cpp
 *
 *  Created on: 29/11/2017
 *      Author: Tato
 */

#include <vector>     //Libreria para manejar vectores.
#include <stdlib.h>   //Libreria para manejar entradas y salidas.
#include <iostream>   //Libreria para manejar entradas y salidas.
#include <string.h>   //Libreria para manejar strings.
using namespace std;  //Declaración de namespace para hacer uso de palabras reservadas.

/*
*Esta función representa el comportamiento de insertionsort, esta recibe
*como parametro un vector con numeros desordenados.
*/
void insertionsort (vector <int> A)
{
	int x = 0;   //Variable para almacenar el valor a ser insertado.
	int index = 0; //Variable para almacenar una posición.
	int length = A.size(); //Variable que almacena el tamaño de el vector.

	for(int i = 0; i < length; i++) //Ciclo que va desde 0 hasta el tamaño del arreglo.
	{
		x = A[i];	
		index = i;

		while(index > 0 && A[index - 1] > x)//Ciclo el cual termina cuando la variable index sea igual a 0 y cuando el vector en la posición index-1 sea igual a la variable x.  
		{
			A[index] = A[index -1]; //Se intercambian las posiciones.
			index = index -1; //Se le resta un numero a la variable index.
		}

		A[index] = x; //El vector en la posición index se cambia por el valor en la variable x.
	}
	
	for(int i = 0; i < length; i++) //Ciclo que va desde 0 hasta el valor que tenga la variable length.
	{
		cout << A[i] << " "; //Se hace la impresión de el array ya ordenado.
	}
}

int main (){

	vector <int> A; //Vector para almacenar numeros desordenados.
	int z = 0; 	
	int tam = 0; 	//Variable que guarda el tamaño de una cadena compuesta de numeros y comas,este va a ser pedido por consola.
	cout << "Cuantos numeros desea organizar?:"; //Se pide el tamaño.
	cin >> tam; //El tamaño se guarda en la variable tam.
	char input[tam]; //Se crea una cadena de caracteres con un tamaño el cual esta guardado en la variable tam.
	char* ptr; 
	cout << "Ingrese " << tam << " numeros desordenados y separados por comas:"; // Se pide que se ingrese un char con datos separados por comas
	cin >> input; //Se guarda en una variable.
	ptr = strtok(input, ","); //Se llama la función strtok para que divida la la cadena cada vez que se encuentre una coma. 

	while(ptr != NULL)
	{
		A[z++] = atoi(ptr); //Cada char se comvierte en entero con la función atoi y se guarda en un vector.
		ptr = strtok(NULL, ","); 
	}

	insertionsort(A); //Se llama la función bubblesort pasandole como parametro el vector de numeros desorganizados.
	return 0;
}

File: BinarySearch/Logic.py
'''
Created on 30/11/2017

@author: Tato
'''

def binarySearch (board,n):

    lowerbound = 0    #Variable la cual almacena la primera posición del vector
    upperbound = len(board)-1  #Variable la cual almacena la ultima posición del vector
    index = -1  """Variable la cual almacena el primer index del vector"""

    while(lowerbound < upperbound): """Ciclo el cual controla que la variable lowerbound no sea mayor a la variable upperbound"""
        middlepoint = (lowerbound + upperbound) / 2 """Variable la cual almacena la posición de la mitad del vector"""
        if (n == board[middlepoint]): """ Condicional que se vuelve true si la variable n pasada como parametro es igual a el dato de el vector en su posición middlepoint"""
            index = middlepoint """ A la variable index se le pasa el valor de la variable middlepoint"""
            break """ Se rompe el while si la condición es true"""
        else:
            if (n < board[middlepoint]): """ Condicional que se vuelve true si la variable n es menor a el dato que este en el vector en su posición middlepoint"""
                upperbound = middlepoint-1 """ Si la condición es true se le asigna a la variable upperbound el valor de la variable middlepoint menos uno"""
            else:
                lowerbound = middlepoint+1 """Si lo condición es false se le asigna a la variable lowerbound el valor de la variable middlepoint mas uno"""

    if (lowerbound == upperbound and board[lowerbound] == n): """Condicional que se vuelve true si el valor de la variable lowerbound es igual a el valor variable upperbound y el valor del vector en la posición lowebound es igual a la variable n pasada como parametro"""
        index = lowerbound

    print ("el numero "+str(n)+" se encuentra en la posicion "+str(index))   """Imprime el valor de la variable index"""
    return index

a = [2, 90, 1, 0, 12, 23, 28, 12, 80, 76, 4, 123, 10000]

print(binarySearch(a, 23))
